### PR TITLE
Only execute epJSON schema generation steps when dependencies are modified

### LINKED ISSUE #10053: epJSON schema generation build steps executed even when dependencies don't change
Issue overview
--------------
Some epJSON schema build steps are always executed even when the dependencies aren't changing.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by nealkruis**
Pull request overview
---------------------
 - Fixes #10053

The cause of the rebuild was the use of [`execute_process`](https://cmake.org/cmake/help/latest/command/execute_process.html?highlight=execute_process) (which "runs commands while CMake is configuring the project, prior to build system generation") and [`add_custom_target`](https://cmake.org/cmake/help/latest/command/add_custom_target.html#command:add_custom_target) (which "has no output file and is always considered out of date"). The only way to set up build-time dependencies is through [`add_custom_command`](https://cmake.org/cmake/help/latest/command/add_custom_command.html?highlight=add_custom_command), which can "specif[y] any output of the custom command as a source file is given a rule to generate the file using the command at build time".

Also, important here is chaining [file dependencies that are used by multiple targets](https://cmake.org/cmake/help/latest/command/add_custom_command.html?highlight=add_custom_command#example-generating-files-for-multiple-targets). Each file dependency in the chain needs to have a "dummy" custom target (not just the last file in the chain).

### Dependency Walkthrough

1. `Energy+.idd` is created during generation using `configure_file` with `Energy+.idd.in`.
2. `Energy+.schema.epJSON` is created during build time using `add_custom_command` that calls `generate_epJSON_schema.py`, and depends on `Energy+.idd`.
3. The target, `generate_embeddable_epJSON_schema`, is a C++ program that reads `Energy+.schema.epJSON` and prints the binary blob of the schema to stdout.
4. `EmbeddedEpJSONSchema.cc` is created during build time using `add_custom_command` that calls a CMake script that 1) calls `generate_embeddable_epJSON_schema` and stores its output in a CMake variable that is used to 2) configure `EmbeddedEpJSONSchema.in.cc`
5. The `energyplus` target depends on `EmbeddedEpJSONSchema.cc`.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: CMakeLists.txt
Comment (Myoldmopar) [inline CMakeLists.txt]: OK, so the new custom command to generate the epJSON schema file will only be run if the epJSON schema is not generated, or if the dependencies are modified.  Excellent.

File: src/EnergyPlus/CMakeLists.txt
Comment (Myoldmopar) [inline src/EnergyPlus/CMakeLists.txt:722]: And this custom command will only run if the embedded schema file is missing or if the epJSON schema command is run again.  Which only happens if the IDD or whatever are modified.  :+1: :+1:

**Review by Myoldmopar — APPROVED** [review]
This is wonderful.  Fixing a paper cut that has been around for a long time.  So the dependency chain is: 
 - make all / build solution includes the E+ target build
 - The E+ target includes a dependency on the built embedded file
 - The command to generate the embedded file has a dependency on the schema generation command
 - The schema generation command has dependencies on the IDD and the generation script, so only changes in those will cause a rebuild

Awesome!  Assuming CI is happy this can merge at will.

**Comment by Myoldmopar** [comment]
@nealkruis did you see that CI already failed.  Maybe a typo?

**Comment by nealkruis** [comment]
Yeah...it's related to a generated file that wasn't cleared when I was testing locally. I'll fix that and push again.

**Comment by nealkruis** [comment]
It turns out a custom command can only have a file dependency if that file is generated by another custom command in the same CMakeLists.txt. I moved the generation of the epJSON schema into the same CMakeLists.txt as the generation of the embedded schema source file.

I also removed the generation of Energy+.schema.epJSON.in since it appeared to be an unnecessary intermediate file. @mbadams5 let me know if there is a purpose for this file that I'm not seeing.

@Myoldmopar, there was also a change to the readthedocs sphinx configuration where it was reading Energy+.schema.epJSON.in and I changed that to read Energy+.schema.epJSON instead. I didn't do any specific testing to make sure this was working correctly. Hopefully the CI will cover this?

I'll keep an eye on the CI in the meantime.

File: .gitignore
Comment (Myoldmopar) [inline .gitignore:52]: Nice.

File: src/EnergyPlus/CMakeLists.txt
Comment (Myoldmopar) [inline src/EnergyPlus/CMakeLists.txt:706]: This line is going to call generate_schema with `"${CMAKE_RUNTIME_OUTPUT_DIRECTORY}"` as the only argv.  This is fine because the current IDD will have been copied in there by this time.  So then generated schema should find the IDD in there and create an EpJSON schema in there as well.

File: scripts/dev/generate_epJSON_schema/generate_epJSON_schema.py
Comment (Myoldmopar) [inline scripts/dev/generate_epJSON_schema/generate_epJSON_schema.py]: And here inside generate_schema, the only argument is the directory to operate inside.  It finds the IDD in there, and creates the schema file in there.

File: doc/readthedocs/sphinx/conf.py
Comment (Myoldmopar) [inline doc/readthedocs/sphinx/conf.py]: But up here, it looks like this is going to be trying to find the generated schema in the repository root.  I think this needs a tweak but I'll try it first.

**Review by Myoldmopar — COMMENTED** [review]
This is an awesome change, I need to check the RTD stuff, but once that's worked out, now that CI is happy I think this can merge in.

**Comment by Myoldmopar** [comment]
@nealkruis OK, got it worked out.  When it builds on RTD, it does not do any CMake-related stuff, so it never produces an "Energy+.idd" file.  It just runs off of the templatized Energy+.idd.in file.  So to accommodate this, I just modified the schema generation script to fallback to finding the ".in" file if it can't find the actual generated IDD file.  This will have no effect on anyone, but if you build the sphinx docs locally, the dummy schema will be in the IDD folder.  So I added it to the git ignore and this is good to go.  I'll do a quick run on RTD and if it's happy, this will merge later.  

Thanks for taking care of this dev paper cut!

**Comment by Myoldmopar** [comment]
FWIW, RTD is Happy https://readthedocs.org/projects/energyplus/builds/20924757/

I think I'll let GHA and Decent also confirm that I didn't break anything with that last change and then merge if everything is green.

**Comment by mbadams5** [comment]
So I'm looking through the changes now.
I am a little confused since I had reworked this to not run as often. The CMake configuration *should* only be running when the IDD changes.

**Comment by Myoldmopar** [comment]
> The CMake configuration _should_ only be running when the IDD changes.

I agree, but it was definitely running like each time I built E+.

@nealkruis I just tested modifying the IDD and building again, and now I get this: 

```
CMake Error at /eplus/repos/3eplus/scripts/dev/generate_embeddable_epJSON_schema/generate_embedded_epJSON_schema.cmake:5 (configure_file):
  No such file or directory
```

Let me know if I can provide anything else.

**Comment by Myoldmopar** [comment]
I made a comment [above](https://github.com/NREL/EnergyPlus/pull/10054#pullrequestreview-1456019376) where I tried to capture the dependency chain.  Perhaps it would be useful to expand on that and describe what files are in place in the raw repository, then what files get generated at configure time, and why.  I basically understand the whole thing, but admittedly, when I have to mentally process the epjson schema generation and the embeddable epjson schema projects, I have to get out my note sheet to keep track of the dependencies of each  I think this is a comment for @nealkruis but @mbadams5 if you care to expand or comment, that's great too.

**Comment by mbadams5** [comment]
The Version object and the fields in the screenshot are not getting configured at compile time anymore. This is kind of odd since I thought you were passing in the configured IDD file... (this is on a completely fresh build fyi). 
Screenshot 2023-06-06 at 2 00 42 PM

**Comment by mbadams5** [comment]
I see now why I thought I had already fixed this issue... because I did on a branch that didn't end up going anywhere or have a pull request. See this commit: https://github.com/NREL/EnergyPlus/commit/03c3e02602ba47a9f5bf056caf5cd3f88fda78d1

**Comment by mbadams5** [comment]
Looking back at my code in that commit, I would highly recommend cherry picking that commit over and using it.
I keeps the timeout for the epJSON code, tracks all the dependency files correctly, and only runs when the IDD file changes.

**Comment by nealkruis** [comment]
> The Version object and the fields in the screenshot are not getting configured at compile time anymore. This is kind of odd since I thought you were passing in the configured IDD file... (this is on a completely fresh build fyi). Screenshot 2023-06-06 at 2 00 42 PM

Good catch. The Python script is changing the version back to the CMake variables. I think we can just remove that step and it should come through from the configured IDD.

File: src/EnergyPlus/CMakeLists.txt
Comment (Myoldmopar) [inline src/EnergyPlus/CMakeLists.txt:718]: Yes!! This looks right!  Testing it out now.

**Comment by Myoldmopar** [comment]
This is working perfectly for me now, thanks @nealkruis !  Having the schema generation as a target makes good sense.  I'm going to let Decent do a pass over this and I'll also make sure RTD is building it properly (it should be).  Then this will merge later or tomorrow if everything is happy.

File: CMakeLists.txt
Comment (mbadams5) [inline CMakeLists.txt:308]: Why move the `configure_file` command for `Energy+.idd.in` into the EnergyPlus CMakeLists? I mean, it doesn't really matter since the EnergyPlus CMakeLists is always going to be run, but there are still other IDD `configure_file` commands here too.

File: CMakeLists.txt
Comment (mbadams5) [inline CMakeLists.txt:307]: The timeout check is missing from your new changes. This is important since if the IDD is malformed, the IDD parser is likely to get into an infinite loop. In my commit, I ended up changing [generate_epJSON_schema.py](scripts/dev/generate_epJSON_schema/generate_epJSON_schema.py) in my previously linked commit to include the timeout feature within that code. I also added a CLI functionality to that script now, which could be used by RTD to call the correct IDD and output the epJSON to a specific spot.

File: src/EnergyPlus/CMakeLists.txt
Comment (mbadams5) [inline src/EnergyPlus/CMakeLists.txt]: Looking at [CMake's add_custom_command documentation](https://cmake.org/cmake/help/latest/command/add_custom_command.html), you should probably add `MAIN_DEPENDENCY`, `BYPRODUCTS`, and `VERBATIM`. Look at [my version here that includes those keywords as well as a `COMMENT` that is displayed during build time](https://github.com/NREL/EnergyPlus/commit/03c3e02602ba47a9f5bf056caf5cd3f88fda78d1#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR301-R304). `BYPRODUCTS` is important for Ninja builds especially, but all builds will properly handle the generated `Energy+.schema.epJSON` since 

> each byproduct file will be marked with the [GENERATED](https://cmake.org/cmake/help/latest/prop_sf/GENERATED.html#prop_sf:GENERATED) source file property automatically

The `MAIN_DEPENDENCY` keyword is important because:

> Specify the primary input source file to the command. This is treated just like any value given to the DEPENDS option but also suggests to Visual Studio generators where to hang the custom command.

File: src/EnergyPlus/CMakeLists.txt
Comment (mbadams5) [inline src/EnergyPlus/CMakeLists.txt]: Similar comment to the other `add_custom_command`. This should include `BYPRODUCTS`, `VERBATIM`, and maybe `COMMENT`.
See [this code as an example](https://github.com/NREL/EnergyPlus/commit/03c3e02602ba47a9f5bf056caf5cd3f88fda78d1#diff-55856ea372630af1ffb24adf007a830b73b57f27bfbd5783767e425df75c59aeR28-R31).

File: src/EnergyPlus/CMakeLists.txt
Comment (mbadams5) [inline src/EnergyPlus/CMakeLists.txt:718]: This target needs to be added [to the `generate_embeddable_epJSON_schema` CMakeLists](https://github.com/NREL/EnergyPlus/blob/10053-epJSON-build-dependencies/scripts/dev/generate_embeddable_epJSON_schema/CMakeLists.txt#L3-L4).

` add_dependencies(generate_embeddable_epJSON_schema epjson_schema)`

`generate_embeddable_epJSON_schema` depends on `Energy+.schema.epJSON` as an input.

**Review by mbadams5 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (nealkruis) [reply CMakeLists.txt:308]: This can be moved back since we have to use dummy targets to establish the file dependency properly. But I wonder if all of this should happen in the idd subdirectory.

File: CMakeLists.txt
Comment (nealkruis) [reply CMakeLists.txt:307]: I feel that this should be addressed/checked in the parser itself, and not be based on elapsed time. I have had occasions where the timeout triggers prematurely. I think another PR with changes to the python script would be welcome, but it isn't really related to the dependency issues addressed here.

File: src/EnergyPlus/CMakeLists.txt
Comment (nealkruis) [reply src/EnergyPlus/CMakeLists.txt:718]: The `generate_embeddable_epJSON_schema` executable does not depend on `Energy+.schema.epJSON`.

`Energy+.schema.epJSON` is only a a dependency of the custom command that calls `generate_embeddable_epJSON_schema`.

File: src/EnergyPlus/CMakeLists.txt
Comment (nealkruis) [reply src/EnergyPlus/CMakeLists.txt]: 1. I'm pretty sure `BYPRODUCTS` is only needed if there are multiple files created by the command. In this case, it is not necessary since there is only a single output.
2. `VERBATIM` should probably be added to all our custom commands (not just the ones we're adding here), but none of them currently have characters that would need to be escaped by CMake.
3. I'll assign a `MAIN_DEPENDENCY`, but it won't have any meaningful impact on the build process since the command will check all dependencies (and not just the one the command is associated with in Visual Studio).
4. `COMMENT` seems redundant since the build system already prints messages when the corresponding dummy targets are built.

File: src/EnergyPlus/CMakeLists.txt
Comment (mbadams5) [reply src/EnergyPlus/CMakeLists.txt:718]: Yeah, you are correct. `Energy+.schema.epJSON` needs to be passed into that executable but isn't needed to compile that executable.

File: CMakeLists.txt
Comment (mbadams5) [reply CMakeLists.txt:307]: In my commit, I did change it to address/check in the parser itself (but still based on elapsed time).
It could be another PR with the changes to the python script.

File: src/EnergyPlus/CMakeLists.txt
Comment (mbadams5) [reply src/EnergyPlus/CMakeLists.txt]: I think `BYPRODUCTS` is still useful and doesn't hurt anything:

> DEPENDS:
New in version 3.16: A target-level dependency is added if any dependency is listed as BYPRODUCTS of a target or any of its build events in the same directory to ensure the byproducts will be available.

> BYPRODUCTS:
Explicit specification of byproducts is supported by the [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja) generator to tell the ninja build tool how to regenerate byproducts when they are missing. It is also useful when other build rules (e.g. custom commands) depend on the byproducts.
The [Makefile Generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#makefile-generators) will remove BYPRODUCTS and other [GENERATED](https://cmake.org/cmake/help/latest/prop_sf/GENERATED.html#prop_sf:GENERATED) files during make clean.

`COMMENT` is a little redundant BUT it does provide a more descriptive comment in the build messages such as `        COMMENT "Generating Energy+.schema.epJSON from Energy+.idd"`

**Comment by Myoldmopar** [comment]
This is working very happily on my machine now.  I can edit the IDD or not and it is building/rebuilding nicely.  Decent CI is also fully happy.  ReadTheDocs is also happy.  I realize there is still a little work that could be done on the script timeout stuff for a follow-up PR, but I think this does fix the fundamental thing trying to be fixed here, which is the dependency cleanup.  @mbadams5 do you want me to pull in your timeout changes into this branch prior to merge?  Or is it OK in a follow-up PR?  And do you have any final comments that need to be addressed before this merges?  Thanks to both of you.

**Comment by Myoldmopar** [comment]
@mbadams5 if you could provide any final comment here, or if you feel this is reasonable for now, at least mark it as approved, that would be great.  I am on board with a follow-up PR to get the timeout stuff implemented again.  If you disagree and think this needs more, feel free to say that as well of course :)

**Comment by Myoldmopar** [comment]
Alright, I'm gonna go ahead and get this merged in.  Everything is happy when I try to edit or IDD or not, it's not rebuilding unnecessarily.  And Decent is happy.  Thanks @nealkruis.  And @mbadams5 if you want to do any follow-up, or want me to do it, just let me know.

**Comment by Myoldmopar** [comment]
So...devs have started having issues where the IDD changes are not triggering schema rebuilds.  I'm not completely sure of the context/situation that results in this, but @RKStrand is currently encountering this, and it's holding up work.  I know @amirroth was seeing this and it was causing unit tests to fail.  I'm possibly reverting this PR to get builds working again and we can try to fix the schema rebuild thing later.

**Comment by nealkruis** [comment]
That's not good. If anyone can reproduce a scenario where it's not rebuilding, I can try to hop on a call to debug the situation.

**Comment by RKStrand** [comment]
You can reproduce it using my branch (8817Kinetic...).  I tried to add two new input fields at the end of the Kinetic battery input.  It can be tested with the PVwithBattery.idf.  If you add the new fields at the end of the battery object, E+ says that the object is not extensible and stops with a severe/fatal error.  In case you need to know, I'm working on a 2019 MacBook Pro that is running MacOS Ventura 13.4.1 (22F82), CMake 3.20.5, and Xcode 14.3.1.  Hope this helps!

**Pull request merged**
